### PR TITLE
Changed the association setup for ip/port

### PR DESCRIPTION
**Pull request by roaet**
Fixes JIRA:NCP-1608
Maybe fixes JIRA:NCP-1613 and JIRA:NCP-1612

### CODE REVIEW

File: quark/db/models.py [hunk 1]
```diff
@@ -170,6 +172,31 @@ class IPAddress(BASEV2, models.HasId):
...
                                nullable=True)
 
+    def is_shared(self):
+        return self.address_type == ip_types.SHARED
+
+    def has_shared_owner(self):
+        for assoc in self["associations"]:
+            if assoc.service != 'none' and assoc.service is not None:
+                return True
+        return False
```
Comment (Cerberus98) [inline quark/db/models.py]: nit: why wouldn't this call get_shared_owner and then simply return True or False under the same conditions?

File: quark/db/models.py [hunk 2]
```diff
@@ -415,7 +441,7 @@ def ip_addresses(cls):
         return orm.relationship(IPAddress, primaryjoin=primaryjoin,
                                 secondaryjoin=secondaryjoin,
                                 secondary=port_ip_association_table,
-                                backref="ports",
+                                backref='ports',
```
Comment (Cerberus98) [inline quark/db/models.py]: why? :snowman:

File: quark/plugin_modules/ip_addresses.py [hunk 3]
```diff
@@ -339,19 +333,19 @@ def update_port_for_ip_address(context, ip_id, id, port):
...
-    addr = db_api.ip_address_find(context, id=ip_id, scope=db_api.ONE)
     sanitize_list = ['service']
-    port_dict = {k: port['port'][k] for k in sanitize_list}
+    addr = db_api.ip_address_find(context, id=ip_id, scope=db_api.ONE)
     if not addr:
         raise quark_exceptions.IpAddressNotFound(addr_id=ip_id)
+    port_db = db_api.port_find(context, id=id, scope=db_api.ONE)
+    if not port_db:
+        raise quark_exceptions.PortNotFound(port_id=id)
+    port_dict = {k: port['port'][k] for k in sanitize_list}
```
Comment (Cerberus98) [inline quark/plugin_modules/ip_addresses.py]: yay, our first real non-python 2.6 compatible code!

File: quark/db/models.py (hunk 1 above)
Comment (roaet) [inline quark/db/models.py]: Being shared and having an owner are different states.

File: quark/db/models.py (hunk 2 above)
Comment (roaet) [inline quark/db/models.py]: Because :cry:

File: quark/plugin_modules/ip_addresses.py (hunk 3 above)
Comment (roaet) [inline quark/plugin_modules/ip_addresses.py]: Is this ok?

File: quark/plugin_views.py [hunk 4]
```diff
@@ -155,10 +155,11 @@ def _make_security_group_rule_dict(security_rule, fields=None):
     return res
 
 
-def _ip_port_dict(port, fields=None):
+def _ip_port_dict(ip, port, fields=None):
+    service = ip.get_service_for_port(port)
     res = {"id": port.get("id"),
            "device_id": port.get("device_id"),
-           "service": port.get("service")}
+           "service": service}
```
Comment (Cerberus98) [inline quark/plugin_views.py]: Worth knowing: unless the relationship is eager loaded, fetching "service" here will likely hit the database to get all associations per port returned in the list (one per access).

File: quark/plugin_modules/ip_addresses.py (hunk 3 above)
Comment (Cerberus98) [inline quark/plugin_modules/ip_addresses.py]: yiiissssss

**Comment by Cerberus98** [comment]
:eggplant: :hankey: :+1:

**Pull request merged**